src/Data/Primitive/Unlifted/Array/ST.hs

```diff
@@ -0,0 +1,497 @@
+  case unST m s of { (# s', MutableUnliftedArray mary# #) ->
+  unsafeFreezeUnliftedArray# mary# s'} of (# _, ary# #) -> ary#
+{-# INLINE runUnliftedArray# #-}
+
+unST :: ST s a -> State# s -> (# State# s, a #)
+unST (ST f) = f
+
+unsafeCreateUnliftedArray
+  :: Int
+  -> (forall s. MutableUnliftedArray s a -> ST s ())
+  -> UnliftedArray a
+unsafeCreateUnliftedArray !n f = runUnliftedArray $ do
+  mary <- unsafeNewUnliftedArray n
+  f mary
+  pure mary
+
+-- | Creates a new 'MutableUnliftedArray'. This function is unsafe because it
+-- initializes all elements of the array as pointers to the empty array. Attempting
+-- to read one of these elements before writing to it is in effect an unsafe
+-- coercion from @'UnliftedArray' a@ to the element type.
+unsafeNewUnliftedArray
+  :: Int -- ^ size
+  -> ST s (MutableUnliftedArray s a)
+{-# inline unsafeNewUnliftedArray #-}
+unsafeNewUnliftedArray (I# i) = ST $ \s -> case unsafeNewUnliftedArray# i s of
+  (# s', ma #) -> (# s', MutableUnliftedArray ma #)
+
+
+-- | Creates a copy of a portion of an 'UnliftedArray'
+cloneUnliftedArray
+  :: UnliftedArray a -- ^ source
+  -> Int -- ^ offset
+  -> Int -- ^ length
+  -> UnliftedArray a
+{-# inline cloneUnliftedArray #-}
+cloneUnliftedArray (UnliftedArray ary) (I# off) (I# len)
+  = UnliftedArray (cloneUnliftedArray# ary off len)
+
+-- | Creates a new 'MutableUnliftedArray' containing a copy of a portion of
+-- another mutable array.
+cloneMutableUnliftedArray
+  :: MutableUnliftedArray s a -- ^ source
+  -> Int -- ^ offset
+  -> Int -- ^ length
+  -> ST s (MutableUnliftedArray s a)
+{-# inline cloneMutableUnliftedArray #-}
+cloneMutableUnliftedArray (MutableUnliftedArray mary) (I# off) (I# len)
+  = ST $ \s -> case cloneMutableUnliftedArray# mary off len s of
+      (# s', mary' #) -> (# s', MutableUnliftedArray mary' #)
+
+emptyUnliftedArray :: UnliftedArray a
+emptyUnliftedArray = UnliftedArray (emptyUnliftedArray# Exts.void#)
+
+singletonUnliftedArray :: PrimUnlifted a => a -> UnliftedArray a
+{-# INLINE singletonUnliftedArray #-}
+singletonUnliftedArray x = runUnliftedArray $ newUnliftedArray 1 x
+
+concatUnliftedArray :: UnliftedArray a -> UnliftedArray a -> UnliftedArray a
+{-# INLINE concatUnliftedArray #-}
+concatUnliftedArray (UnliftedArray a1) (UnliftedArray a2)
+  = UnliftedArray (concatUnliftedArray# a1 a2)
+
+-- This junk is to make sure we unbox properly. Inlining this doesn't seem
+-- likely to be much of a win ever, and could potentially lead to reboxing,
+-- so we NOINLINE. It would be nice to find a prettier way to do this.
+concatUnliftedArray# :: UnliftedArray# a -> UnliftedArray# a -> UnliftedArray# a
+{-# NOINLINE concatUnliftedArray# #-}
+concatUnliftedArray# a1 a2 =
+  let !sza1 = sizeofUnliftedArray# a1
+  in
+    if Exts.isTrue# (sza1 Exts.==# 0#)
+    then a2
+    else
+      let !sza2 = sizeofUnliftedArray# a2
+      in
+        if Exts.isTrue# (sza2 Exts.==# 0#)
+        then a1
+        else Exts.runRW# $ \s ->
```

Comment:
Should we `noDuplicate#` here to make sure we don't concatenate in multiple threads? Should we do so only if the final array size exceeds some value?